sw/device/lib/dif/dif_spi_device.c

```diff
@@ -0,0 +1,214 @@
+  } else {
+    message_len = space_left;
+  }
+
+  size_t bytes_left = message_len;
+
+  // First, bring txf_write into word-alignment, so we can do full-word writes.
+  if (!IS_WORD_ALIGNED(txf_write_addr)) {
+    size_t misalignment = txf_write_addr & 0b11;
+    size_t needed_bytes = sizeof(uint32_t) - misalignment;
+    if (needed_bytes > bytes_left) {
+      needed_bytes = bytes_left;
+    }
+
+    uint32_t value =
+        reg32_read(base_addr, SPI_DEVICE_TXF_OFFSET(txf_write_addr));
+    __builtin_memcpy(((void *)&value) + misalignment, data, needed_bytes);
+    reg32_write(base_addr, SPI_DEVICE_TXF_OFFSET(txf_write_addr), value);
+
+    fifo_increment(&txf_write_addr, needed_bytes, SPI_DEVICE_TXF_SIZE);
+    data += needed_bytes;
+    bytes_left -= needed_bytes;
+  }
+
+  // Now we can treat txf_write_addr as being word-aligned, assuming we have
+  // data left to write.
+  while (bytes_left != 0) {
+    size_t bytes_to_copy = bytes_left < 4 ? bytes_left : 4;
+
+    uint32_t value = 0;
+    __builtin_memcpy(((void *)&value), data, bytes_to_copy);
+    reg32_write(base_addr, SPI_DEVICE_TXF_OFFSET(txf_write_addr), value);
+
+    fifo_increment(&txf_write_addr, bytes_to_copy, SPI_DEVICE_TXF_SIZE);
+    data += bytes_to_copy;
+    bytes_left -= bytes_to_copy;
+  }
+
+  txf_ptr = txf_write_addr << SPI_DEVICE_TXF_PTR_WPTR_OFFSET |
+            txf_read_addr << SPI_DEVICE_TXF_PTR_RPTR_OFFSET;
+  reg32_write(base_addr, SPI_DEVICE_TXF_PTR_REG, txf_ptr);
+  return message_len;
+}
+
+size_t spi_device_recv(reg32_t base_addr, void *buf, size_t len) {
+  uint32_t rxf_ptr = reg32_read(base_addr, SPI_DEVICE_RXF_PTR_REG);
+  uint32_t rxf_write_addr = (rxf_ptr >> SPI_DEVICE_RXF_PTR_WPTR_OFFSET) &
+                            SPI_DEVICE_RXF_PTR_WPTR_MASK;
+  uint32_t rxf_read_addr = (rxf_ptr >> SPI_DEVICE_RXF_PTR_RPTR_OFFSET) &
+                           SPI_DEVICE_RXF_PTR_RPTR_MASK;
+
+  size_t fifo_in_use_len =
+      fifo_depth(rxf_write_addr, rxf_read_addr, SPI_DEVICE_RXF_SIZE);
+
+  if (fifo_in_use_len == 0) {
+    return 0;
+  }
+
+  size_t message_len = fifo_in_use_len > len ? len : fifo_in_use_len;
+  size_t bytes_left = message_len;
+
+  // First, bring rxf_read into word alignment, so that we can do full-word
+  // reads later.
+  if (!IS_WORD_ALIGNED(rxf_read_addr)) {
+    size_t misalignment = rxf_read_addr & 0b11;
+    size_t needed_bytes = sizeof(uint32_t) - misalignment;
+    if (needed_bytes > bytes_left) {
+      needed_bytes = bytes_left;
+    }
+
+    uint32_t value =
+        reg32_read(base_addr, SPI_DEVICE_RXF_OFFSET(rxf_read_addr));
+    __builtin_memcpy(buf, ((void *)&value) + misalignment, needed_bytes);
+
+    fifo_increment(&rxf_write_addr, needed_bytes, SPI_DEVICE_RXF_SIZE);
+    buf += needed_bytes;
+    bytes_left -= needed_bytes;
+  }
+
+  // Now we can just treat rxf_read as being word-aligned, if we still have data
+  // left.
+  while (bytes_left != 0) {
+    size_t bytes_to_copy = bytes_left < 4 ? bytes_left : 4;
```

Comment:
A define for magic number "4", or a comment describing what it represents? I understand that in this particular case its pretty clear what it is, but just in general for consistency.

Reply:
It's `sizeof(uint32_t)`, which is what I've replaced it with everywhere.